### PR TITLE
initial implementation of spa

### PR DESCRIPTION
**Pull request by ahwillia**
Here is an initial implementation of SPA. See reference here: http://arxiv.org/abs/1208.1237

Happy to iterate on this and improve the code as you see fit.

### CODE REVIEW

**Comment by silenj** [comment]
Hi, seems the spa fails for non square matrices. The spa function update projections have to be ordered differently R -= p*p' * R / norm(p)^2 . with this change the results are reasonable. I have yet not made much to look at details but would suggest spa to be included with NMF package.

**Comment by ahwillia** [comment]
@silenj -- thanks, the commit should fix this.

@lindahua -- is there a way to get this looked at?

**Comment by silenj** [comment]
I do not think there are formal ways. I compared decompositions of a few hundred spectra each with 30 kbins. The results are similar but maybe more reasonable than ordinary nmf algorithms.

**Comment by ahwillia** [comment]
@silenj -- are those datasets separable or approximately separable (see definition in here: http://arxiv.org/pdf/1208.1237v3.pdf)? Is it much noticeably quicker than the other algorithms?

**Comment by kmsquire** [comment]
(Open and closed to rerun CI.)

File: src/interf.jl [hunk 1]
```diff
@@ -6,7 +6,7 @@ function nnmf{T}(X::AbstractMatrix{T}, k::Integer;
                  maxiter::Integer=100,
                  tol::Real=cbrt(eps(T)/100),
                  verbose::Bool=false)
-
+
```
Comment (kmsquire) [inline src/interf.jl]: It's preferable not to introduce whitespace.

File: src/spa.jl [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	function SPA(;obj::Symbol=:mse)
+		obj == :mse || obj == :div || error("Invalid value for obj.")
+        new(obj)
+    end
+end
+
+# initialization
+function spa(
+	X::Matrix,
+	k::Int;
```
Comment (kmsquire) [inline src/spa.jl]: Could be relaxed to `Integer`.

File: src/spa.jl [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	nnls_alg = (:pivot,:cache)
+	)
+
+	# Normalize data so that columns of X sum to one
+	R = X ./ sum(X,1)
+
+	# W = R[:,ai], where ai are the "anchor indices"
+	# (ai forms the convex hull of columns in R)
+	ai = (Int64)[]
```
Comment (kmsquire) [inline src/spa.jl]: And this could just be `Int`.

**Comment by ahwillia** [comment]
Thanks @silenj @kmsquire -- fixes are now in the latest commit.

**Comment by kmsquire** [comment]
Thanks @ahwillia!  Can you also please add some tests (in `NMF/test`) similar to the ones there, and squash the commits?  Assuming all looks well, I'll merge after that.

**Comment by silenj** [comment]
Great! Thanks for this quick reaction. Next we need to put in some brief 
documents
into the package docs to finalize this.

Regards Johan

On 06.02.2016 21:01, Kevin Squire wrote:
>
> Thanks @ahwillia <https://github.com/ahwillia>! Can you also please 
> add some tests (in |NMF/test|) similar to the ones there, and squash 
> the commits? Assuming all looks well, I'll merge after that.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/JuliaStats/NMF.jl/pull/12#issuecomment-180838994>.
>

-- 
Private:  Johan Silen, +358-505423872

**Comment by ararslan** [comment]
If you're still interested in this PR, could you please rebase your branch and resolve conflicts?

**Comment by ahwillia** [comment]
Sorry that this is abandoned. Its basically done, but last time I checked I wasn't happy with the result for some reason. Also still need to add tests and docs.

What are the plans of this repo going forward? If it is going to be more actively maintained, then I could find time to clean this up and merge. Honestly, I would like to give this whole repo a makeover, but it would be difficult without someone in JuliaStats supporting the effort.

**Comment by ararslan** [comment]
I think @lindahua had been the primary maintainer but I haven't seen him around in a while. I can help maintain it from an administrative perspective (reviewing, merging, moral support, etc.) but I don't know a whole lot about the subject matter so I don't know how meaningful my feedback would be.

cc @andreasnoack or @jiahao, two of our resident experts on matrices and stuff

**Comment by tsano430** [comment]
@ahwillia PR #50 was merged. Ok to close?

**Pull request closed without merging**